### DISCUSSION TITLE
Variable substitution is not working

### DISCUSSION THREAD

**Discussion by sw-ha-personal** [fluxcd/flux2#4419] [Q&A]
I have two kustomization, one is parent and child 

Parent Kustomization: ./cluster/my-clusters/flux-system/k.yaml

apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  name: flux-system
  namespace: flux-system
spec:
  interval: 10m0s
  path: ./cluster/my-clusters
  prune: true
  sourceRef:
    kind: GitRepository
    name: flux-system    
  postBuild:
    substitute:
      env: "testing"

Child Kustomization: ./cluster/my-clusters/k.yaml

apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  name: dev-kust
  namespace: flux-system
spec:
  interval: 10m0s
  path: ./cluster/my-clusters/dev
  prune: true
  sourceRef:
    kind: GitRepository
    name: flux-system
  postBuild:
    substitute:
      env: ${env}

I have a pod resource: ./cluster/my-clusters/dev/p.yaml

apiVersion: v1
kind: Pod
metadata:
  name: d
  namespace: dev
  labels:
   env: ${env}
spec:
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
    
    

   After applying above, the pod gets created without any labels.
   "dev-kust" kustomization reconciled successfuly
   
   But Flux-system kustomization says below:
   
  dev-kust        main@sha1:05c0dc5d      False           True    Applied revision: main@sha1:05c0dc5d
  
 flux-system     main@sha1:a1669e1b      False   False   Kustomization/flux-system/dev-kust dry-run failed, reason: Invalid: Kustomization.kustomize.toolkit.fluxcd.io "dev-kust" is invalid: [spec.postBuild.substitute.env: Invalid value: "null": spec.postBuild.substitute.env in body must be of type string: "null"]

**Comment by kingdonb** [comment]
It is hard to read your example, without the blockquotes (```)

Since it depends on indenting. Broadly, I see two things wrong. One, it looks like you have written your own flux-system file and edited it - this is not recommended, we recommend the bootstrap method which creates a file that you should not edit. You can customize it with patches, but you need to limit the customizations to things which will not break bootstrap if you need to run bootstrap again later.

At first I saw it looks like the child kustomization is not applied by the parent kustomization, so it does not impart the `env: testing` value. But now I see your flux-system is meant to be passing a value through and it seems like your dev-kust has been applied with an invalid value, since environment variables cannot be set to null. Did you kubectl apply all the files in this directory? That is not how bootstrap is meant to work, it's meant to flow from a central source with minimal config, out to the others that do the detailed work.

I think something applied the dev-kust without the environment variable being included.

I was confused by your example because it's hard to read , it looks like you might be using the `substitute` method based on the docs here:

https://fluxcd.io/flux/components/kustomize/kustomizations/#post-build-variable-substitution

Can you clean up the YAML examples or provide a public repo with the issue reproduced there, so there is less to do for someone that wants to reproduce the issue you had?

There might actually be some UX issue here, but from what you're showing me, if you bootstrapped Flux properly then I would think the variable would be applied. I can't really see from here if you did or if you didn't, but it looks like you might have gone and applied all of the files with an envsubst that was missing an environment setting and dropped some values. Or it could be a spacing error. I'm not sure if it's supposed to work to pass variables through env like this from one kustomization to another, but I think it should work. I'll give it a try if I can reproduce the issue.

(Did you by any chance already solve the problem, and maybe you can help us close the thread, or are you still struggling with this?)
